client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -214,9 +215,14 @@ private long getRemainingBytes(SegmentMetadataClientFactory metaFactory, StreamC
     }
 
     @Override
-    public Observable<SegmentEvent> getSegmentEventNotifier(ScheduledExecutorService executor) {
+    public Observable<SegmentNotification> getSegmentNotifier(ScheduledExecutorService executor) {
         checkNotNull(executor, "executor");
         return this.notifierFactory.getSegmentNotifier(this::createSynchronizer, executor);
     }
 
+    @Override
+    public Observable<EndOfDataNotification> getEndOfDataNotifier(ScheduledExecutorService executor) {
+        checkNotNull(executor, "executor");
+        return this.notifierFactory.getEndOfDataNotifier(this::createSynchronizer, executor);
```

Comment:
Perhaps rather than a Synchronizer factory, that is re-created and re-read every time, it could just pass in the Synchronizer. They are, after all capable of updating. If the notificationSystem or NotificationFactory was given one when it was constructed, it would only ever need to get the updates by calling fetchUpdates(). I think this model would work better with an internal threadpool though...

Reply:
Done. Refactored the code to ensure a single stateSynchronizer instance is used by the notifiers.